subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy

```diff
@@ -16,34 +16,69 @@
 
 package org.gradle.api.internal.changedetection
 
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.InputDirectory
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.daemon.DaemonLogsAnalyzer
+import org.gradle.test.fixtures.server.http.BlockingHttpServer
+import org.gradle.workers.WorkerExecutor
+import org.junit.Rule
 import spock.lang.Issue
 
 class CorruptedTaskHistoryIntegrationTest extends AbstractIntegrationSpec {
+    @Rule
+    BlockingHttpServer server = new BlockingHttpServer()
 
     // If this test starts to be flaky it points to a problem in the code!
     // See the linked issue.
     @Issue("https://github.com/gradle/gradle/issues/2827")
     def "broken build doesn't corrupt the artifact history"() {
+        server.start()
         def numberOfFiles = 10
         def numberOfOutputFilesPerTask = numberOfFiles
         def numberOfInputProperties = 10
         def numberOfTasks = 100
         def totalNumberOfOutputDirectories = numberOfTasks
-        def killPollInterval = 10
+        def numberOfExecutedTaskBeforeKilling = 11
         def totalNumberOfOutputFiles = numberOfTasks * numberOfOutputFilesPerTask + totalNumberOfOutputDirectories
 
-        setupTestProject(numberOfFiles, numberOfInputProperties, numberOfTasks, killPollInterval)
+        setupTestProject(numberOfFiles, numberOfInputProperties, numberOfTasks)
 
         executer.beforeExecute {
             // We need a separate JVM in order not to kill the test JVM
             requireGradleDistribution()
+            requireIsolatedDaemons()
+            requireDaemon()
         }
 
         when:
         succeeds("createFiles")
         succeeds("clean")
-        fails("createFiles", "-PkillMe=true", "--max-workers=${numberOfTasks}")
+
+        // Worker gate
+        def workerHandler = server.expectConcurrentAndBlock(numberOfTasks, (1..numberOfTasks).collect { "worker-for-output$it" } as String[])
+
+        // Kill gate
+        def killHandler = server.expectOptionalAndBlock(numberOfExecutedTaskBeforeKilling, (1..numberOfTasks).collect { "kill-ready-for-output$it" } as String[])
+
+        // Start Gradle in the background
+        def handle = executer.withArguments("createFiles", "-DkillMe=true", "--max-workers=${numberOfTasks}").start()
+
+        // Wait for all worker to start
+        workerHandler.waitForAllPendingCalls()
+
+        // Release only a few of them
+        workerHandler.release(numberOfExecutedTaskBeforeKilling)
+
+        // Wait until workers finish their jobs
+        killHandler.waitForAllPendingCalls()
+
+        // Kill the daemon
+        killRunningDaemon()
```

Comment:
❌ The idea is to kill the daemon while some tasks are writing their task history, not after they produced the output files, before the task finishes. So in its current state, the test won't do anything anymore.

Reply:
How do we block mid-way writing the task history?

Reply:
We can't 😞 
The idea of the test was (and it did reproduce the original problem back then), to kill the daemon while enough tasks were writing the task history, so we would sometimes catch one of those tasks. We can achieve the same with blocking http server I guess.